docs/sdk/instructlab-sdk.md

```diff
@@ -0,0 +1,171 @@
+
+1. Setup a custom config file for ilab (optional)
+2. Initialize a taxonomy
+3. Ensure their taxonomy is valid
+4. Ingest some data for RAG and SDG (SDG coming soon)
+5. Generate synthetic data using an InstructLab pipeline
+6. Do some custom handling per their use case
+7. Fine-tune a model using the custom config they initialized for their hardware
+8. Evaluate their model after training using various benchmarks
+
+A user could do the following to accomplish this if they had an SDK:
+
+```python
+ilab_client = InstructLab(config_path="", taxonomy="", auto_detect=True)
+if ilab_client.taxonomy.diff():
+    rag_ingestor = RAG.Ingestor(...)
+    rag_ingestor.ingest()
+
+    data_ingestor = Data.Ingestor()
+    data_ingestor.ingest()
+
+    data_generator = Data.Generator(...)
+    data_paths = data_generator.generate_data(...)
+
+    some_custom_handling(data_paths)
+
+    model_trainer = Model.Trainer(...)
+    # potentially could expose other things besides training
+    training_data = model_trainer.process_data(...)
+    model_path = model_trainer.train_model(training_data=training_data)
+
+    eval_client = Model.Evaluator(model_path=model_path)
+    eval_client.mt_bench(...)
+    eval_client.dk_bench(...)
+```
+
+(the structure of the SDK and actual arguments is discussed below)
+
+However today, users are forced to run a sequence of commands tailored to only work with the proper directory structure on the system.
+
+## Major Goals
+
+1. Modularize the InstructLab workflow such that any part can be run independently
+2. Allow users to choose whether or not to take advantage of the config/system-profile method of running InstructLab. Meaning they do not need any pre-existing configuration to run the SDK.
+3. Standardize user contracts for the existing functionality of the InstructLab workflow. Existing CLI commands should be using the SDK once past click parsing, not separate code.
+4. Define Contracts loose enough that functionality can be expanded as more advanced features are released.
+5. Document SDK usage in upcoming InstructLab releases.
+
+## Non-Goals
+
+1. Exposing all library functionality immediately
+2. Replacing CLI
+3. Shipping an SDK that is generally available as opposed to v1alpha1 or v1beta1.
+
+## Design
+
+### Versioning
+
+The SDK would start at version v1alpha1 such that it can change/break at any time for the first few iterations as libraries adjust their API surface.
+
+### Structure
+
+This SDK should live in a net new package inside of `instructlab/instructlab` preferably to limit unnecessary imports in a new repository. The SDK could be imported as `instructlab.core...`
+
+The user surface initially should look like this:
+
+`instructlab.core` contains all SDK definitions. Users can `from instructlab.core import...` to use specific SDK classes
```

Comment:
Why not just use `sdk` instead of `core`?

Reply:
I agree w/ this naming convention.

Reply:
Nathan and I talked and going w/ core makes more sense given that we aren't following and of the norms of packaging or shipping an SDK with this Python SDK so keeping the naming as loose as possible is probably better. Other projects tend to use `core` for their main functionality they intend to be consumed I have found.

Reply:
Yes, I've come around on `core` :smile: